lib/report/generate.js

```diff
@@ -0,0 +1,159 @@
+ */
+function getAssets(config) {
+  var lines = [];
+  // gather script, link, and style tags
+  config.assets.forEach(function(asset) {
+    if (asset.script) {
+      lines.push('<script src="' + _.escape(asset.script) + '"></script>');
+    }
+    else if (asset.stylesheet) {
+      lines.push('<link rel="stylesheet" href="' + _.escape(asset.stylesheet) + '">');
+    }
+    else if (asset.style) {
+      var cssPath = asset.style.match(/^\.\/components\//) ? path.join(__dirname, asset.style) : asset.style;
+      var css = fileUtils.read(cssPath);
+      lines.push('<style>\n' + css + '\n</style>');
+    }
+    else if (asset.js) {
+      var jsPath = asset.js.match(/^\.\/components\//) ? path.join(__dirname, asset.js) : asset.js;
+      var js = fileUtils.read(jsPath);
+      lines.push('<script>\n' + js + '\n</script>');
+    }
+  });
+  return lines.join('\n');
+}
+
+/**
+ * Take the final html file and populate placeholders
+ * @param {String} template  The base HTML file
+ * @param {Object} config  The wbsm config file values
+ * @param {Object} data  An object containing the replacement values
+ * @property {String} data.content  The content generated by mdToHtml()
+ * @property {String} data.assets  The assets html generated by getAssets()
+ * @property {String} data.version  The wbsm version as stored in version.js
+ * @property {String} data.dateText  The report generation date
+ * @returns {*}
+ */
+function compileHtml(template, config, data) {
+  var report = template;
+  report = report.replace(/{{content}}/, data.content);
+  report = report.replace(/{{assets}}/, data.assets);
+  // stamp the wbsm version into the report
+  report = report.replace(/{{version}}/, data.version);
+  // replace "{{reportDate}}" with a text string of the date and time when generated
+  report = report.replace(/{{reportDate}}/, data.dateText);
+  // replace "{{reportTitle}}" with the value loaded from the config file
+  report = report.replace(/{{reportTitle}}/, config.reportTitle);
+  return report;
+}
+
+// Read the contents of a file out synchronously and return the file contents.
+function generate(markdownFilename, reportFilename) {
+  // console.log("--------------------md file: ", markdownFilename)
+  // read markdown file contents
+  var mdContents = fileUtils.read(markdownFilename);
+  // console.log("--------------------contents", mdContents)
+
+  if (typeof mdContents === 'undefined') {
+    console.error('No markdown contents found!');
+    process.exit(1);
+  }
+
+  var renderTemplateFilename = settings.getTemplateFilename()
+  var dateText = new Date().toLocaleString()
+  var config = JSON.parse(fileUtils.read("./.wbsm-config.json"));
+  applyConfigDefaults(config);
+  var content = mdToHtml(mdContents, config);
+  var assets = getAssets(config);
+  var template = fileUtils.read(renderTemplateFilename);
+  var report = compileHtml(template, config, {version, dateText, assets, content});
+  fileUtils.write(report, reportFilename);
+}
```

Comment:
So lines 128 to 132 are basically the parts that allow plugins. See the function documentation above to see what each one does. Basically we take a minimal wbs_deliverables_layout.html and allow plugins to add assets in `<head>`, html inside `<body>` and plugin to markdown-it. We define three plugins by default. See them in htmlPlugins directory, changesets below.